numpy/_core/src/multiarray/mapping.c

```diff
@@ -2089,11 +2089,17 @@ array_assign_subscript(PyArrayObject *self, PyObject *ind, PyObject *op)
     else {
         /* May need a generic copy function (only for refs and odd sizes) */
         NPY_ARRAYMETHOD_FLAGS transfer_flags;
-        npy_intp itemsize = PyArray_ITEMSIZE(self);
+        PyArrayObject *src, *dst=self;
+        if (!PyDataType_ISLEGACY(PyArray_DESCR(self))) {
+            src = tmp_arr;
+        }
+        else {
+            src = self;
+        }
```

Comment:
The below makes sense to me, but this one I don't understand?  Legacy should work identically, no?

Reply:
I agree it should, in practice lots of tests seem to depend on the old behavior. As I said in the description, I don't fully understand what this code block is doing, it seems to be here to capture oddball conversions (unicode to half goes through here apparently - one of the failing tests I was looking at).

Reply:
In that case you are using the wrong `descr` below, and similar tests must fail also for string dtype (maybe not because there is an object special path that uses refcheck).

Check MapIter, `descr` is passed, it sets up a buffering iterator to keep the copy loop simple, that ensures that the descriptor is identical between src/dst.  It is not for your dtype, so you need to fetch the src one from the iterator.

Reply:
I tried doing it this way, and it almost works, but the dtype array that lives on the iterator contains references to the dtypes that were passed in to create the iterator, so it still doesn't fix the issue for stringdtype. I tried checking for the case when the descriptor gets replaced and then updated the iterator's descriptor array, but for mysterious reasons that led to crashes.

Maybe the rewrite in the latest push is a bit more appealing? The basic difference between the case with stringdtype and the case that's breaking for other dtypes is for stringdtype the MapIter is allocating an array inline, but for other dtypes the `extra_op` is already allocated. I added a flag that marks whether or not the iterator allocated an array and used it here to distinguish between the cases.

If not I'd appreciate it if you could try your hand at a fix, I tried for several hours today to get this right but couldn't manage it along the lines that you outlined the other day.

Reply:
This is better, I agree it's all a bit of a maze although, I want would like to see if we can't untangle it.  I might help to make `descr` (the one used for the extra-op/tmp array) a strong reference.

One thing that would may be better for all paths anyway, may be to use `descr = finalize_descr(descr)`.  That way, we don't pollute the assigned arrays side car buffer with temporary copies!
That will mean that we would need to pass `_NPY_ARRAY_ENSURE_DTYPE_IDENTITY` when creating copies *and* actually make sure we honor that flag (which we should anyway).

Anyway, need to take another look and it may still be subtly wrong in another pass if `REFCHK` is false, although I am not _sure_ that is ever relevant.

Reply:
That is a good idea to use finalize_descrs when setting up the iterator, I might try my hand at that.